protocol/contracts/amo/helpers/PoolHelper.sol

```diff
@@ -0,0 +1,248 @@
+      if (spotPriceScaled > templePriceFloorRatio.numerator + slippageTPF) {
+          return true;
+      }
+      return false;
+    }
+
+     function isSpotPriceAboveTPF() external view returns (bool) {
+        uint256 spotPriceScaled = getSpotPriceScaled();
+        if (spotPriceScaled > templePriceFloorRatio.numerator) {
+            return true;
+        }
+        return false;
+    }
+
+    // @notice will exit take price above tpf by a percentage
+    // percentage in bps
+    // tokensOut: expected min amounts out. for rebalance this is expected Temple tokens out
+    function willExitTakePriceAboveTPFUpperBound(uint256 tokensOut) external view returns (bool) {
+        uint256 percentageIncrease = (templePriceFloorRatio.numerator * rebalancePercentageBoundUp) / templePriceFloorRatio.denominator;
+        uint256 maxNewTpf = percentageIncrease + templePriceFloorRatio.numerator;
+        (, uint256[] memory balances,) = balancerVault.getPoolTokens(balancerPoolId);
+        uint256 stableIndexInPool = templeBalancerPoolIndex == 0 ? 1 : 0;
+        // a ratio of stable balances with quote price and fees
+        uint256 newTempleBalance = balances[templeBalancerPoolIndex] - tokensOut;
+        uint256 quote = (balances[stableIndexInPool] * TPF_PRECISION ) / newTempleBalance;
+        return quote > maxNewTpf;
+    }
+
+    function willJoinTakePriceBelowTPFLowerBound(uint256 tokensIn) external view returns (bool) {
+        uint256 percentageDecrease = (templePriceFloorRatio.numerator * rebalancePercentageBoundLow) / templePriceFloorRatio.denominator;
+        uint256 minNewTpf = templePriceFloorRatio.numerator - percentageDecrease;
+        (, uint256[] memory balances,) = balancerVault.getPoolTokens(balancerPoolId);
+        uint256 stableIndexInPool = templeBalancerPoolIndex == 0 ? 1 : 0;
+        // a ratio of temple balances with quote price and fees
+        uint256 newTempleBalance = balances[templeBalancerPoolIndex] + tokensIn;
+        uint256 spot = (balances[stableIndexInPool] * TPF_PRECISION) / newTempleBalance;
+        return spot < minNewTpf;
+    }
+
+    // get slippage between spot price before and spot price now
+    function getSlippage(uint256 spotPriceBeforeScaled) external view returns (uint256) {
+        uint256 spotPriceNowScaled = getSpotPriceScaled();
+        // taking into account both rebalance up or down
+        uint256 slippageDifference;
+        if (spotPriceNowScaled > spotPriceBeforeScaled) {
+            unchecked {
+                slippageDifference = spotPriceNowScaled - spotPriceBeforeScaled;
+            }
+        } else {
+            unchecked {
+                slippageDifference = spotPriceBeforeScaled - spotPriceNowScaled;
+            }
+        }
+        return (slippageDifference * TPF_PRECISION) / spotPriceBeforeScaled;
+    }
+
+    function getMax(uint256 a, uint256 b) external pure returns (uint256 maxValue) {
+        // if (a < b) {
```

Comment:
Would ternary give the same gas here and save from having to use assembly?

```
return (a < b) ? b : a;
```

Reply:
did some tests on different implementations. turns out assembly still the winner here

```
contract TestFunc {
    uint256 ans;

    // function saveMaxAssembly(uint256 a, uint256 b) external returns (uint256 maxValue) {
    //      assembly {
    //         if lt(a, b) { maxValue := b }
    //         if iszero(maxValue) { maxValue := a }
    //     }
    //     ans = maxValue;
    //     // execution cost 44254 gas
    // }

    function saveMaxAssemblyMstore(uint256 a, uint256 b) external returns (uint256) {
        ans = a;
        assembly {
            mstore(0, a)
            if lt(a, b) { mstore(0, b) }
            return (0, 0x20)
        }
        // execution cost 44055 gas
    }

    // function saveMax(uint256 a, uint256 b) external returns (uint256 maxValue) {
    //     maxValue =  (a < b) ? b : a;
    //     ans = maxValue;

    //     // execution cost 44235 gas
    // }

    // function saveMax(uint256 a, uint256 b) external returns (uint256) {
    //     ans = a;
    //     return (a < b) ? b : a;

    //     // execution cost 44235 gas
    // }

    // function saveMax(uint256 a, uint256 b) external returns (uint256 maxValue) {
    //     ans = a;
    //     return (a < b) ? b : a;
    //     // execution cost 44235 gas
    // }

}

```

Reply:
weird `unchecked` doesn't save much gas here. will look into this later.